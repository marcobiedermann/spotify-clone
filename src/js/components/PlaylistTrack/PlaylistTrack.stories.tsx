import type { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import PlaylistTrack from '.';
import playlistTrackFixtures from './__fixtures__';

const meta: Meta<typeof PlaylistTrack> = {
  component: PlaylistTrack,
  args: playlistTrackFixtures,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <table>
          <tbody>
            <Story />
          </tbody>
        </table>
      </MemoryRouter>
    ),
  ],
};

type Story = StoryObj<typeof PlaylistTrack>;

const Default: Story = {};

export { Default };
export default meta;
