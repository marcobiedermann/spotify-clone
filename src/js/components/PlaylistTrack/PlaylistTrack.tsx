import format from 'date-fns/format';
import React from 'react';
import { Link } from 'react-router-dom';
import Image from '../Image';
import Media, { MediaBody, MediaObject } from '../Media';
import styles from './PlaylistTrack.module.css';

const defaultImageSize = 48;

interface Image {
  height: number | null;
  url: string;
  width: number | null;
}
interface Album {
  images: Image[];
  id: string;
  name: string;
}

interface Owner {
  name?: string;
  id: string;
}

export interface PlaylistTrackProps {
  added_at: string;
  track: {
    album: Album;
    artists: Owner[];
    duration_ms: number;
    id: string;
    name: string;
    track_number: number;
  };
}

function PlaylistTrack(props: PlaylistTrackProps): JSX.Element {
  const { added_at, track } = props;

  return (
    <tr>
      <td>{track.track_number}</td>
      <td>
        <Media>
          <MediaObject>
            <Image
              url={track.album.images[0].url}
              width={defaultImageSize}
              height={defaultImageSize}
              alt={track.album.name}
            />
          </MediaObject>
          <MediaBody>
            <div className={styles.track__name}>{track.name}</div>
            <div>
              {track.artists
                .map<React.ReactNode>((artist) => (
                  <Link key={artist.id} to={`/artists/${artist.id}`}>
                    {artist.name}
                  </Link>
                ))
                .reduce((prev, curr) => [prev, ', ', curr])}
            </div>
          </MediaBody>
        </Media>
      </td>
      <td>
        <Link to={`/albums/${track.album.id}`}>{track.album.name}</Link>
      </td>
      <td>{format(new Date(added_at), 'MMM d, y')}</td>
      <td align="right">{format(track.duration_ms, 'm:ss')}</td>
    </tr>
  );
}

export default PlaylistTrack;
