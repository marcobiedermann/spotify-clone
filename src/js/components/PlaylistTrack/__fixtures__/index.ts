const playlistTrack = {
  added_at: '2015-01-15T12:39:22Z',
  track: {
    album: {
      images: [
        {
          height: 640,
          url: 'https://i.scdn.co/image/ab67616d0000b273ce6d0eef0c1ce77e5f95bbbc',
          width: 640,
        },
      ],
      id: '5Opp8ursJ6XoTTcqElvAJG',
      name: 'Tibetan Bowls',
    },
    artists: [
      {
        id: '5VQE4WOzPu9h3HnGLuBoA6',
        name: 'Vlasta Marek',
      },
    ],
    duration_ms: 730066,
    id: '64IL0ZrOV7F6BDPm5fhfUa',
    name: 'Is',
    track_number: 2,
  },
};

export default playlistTrack;
