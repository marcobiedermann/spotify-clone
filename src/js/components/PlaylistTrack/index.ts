import PlaylistTrack from './PlaylistTrack';

export default PlaylistTrack;
export * from './PlaylistTrack';
