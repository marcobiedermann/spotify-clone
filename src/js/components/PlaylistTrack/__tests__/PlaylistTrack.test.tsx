import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { expect, it } from 'vitest';
import PlaylistTrack from '..';
import playlistTrackFixtures from '../__fixtures__';

it('renders correctly', () => {
  const { container } = render(
    <MemoryRouter>
      <PlaylistTrack {...playlistTrackFixtures} />
    </MemoryRouter>,
  );

  expect(container.firstChild).toMatchSnapshot();
});
