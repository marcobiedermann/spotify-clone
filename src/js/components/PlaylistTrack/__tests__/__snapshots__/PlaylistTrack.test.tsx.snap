// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<tr>
  <td>
    2
  </td>
  <td>
    <div
      class="media"
    >
      <div
        class="media__object media__object--left"
      >
        <img
          alt="Tibetan Bowls"
          height="48"
          src="https://i.scdn.co/image/ab67616d0000b273ce6d0eef0c1ce77e5f95bbbc"
          width="48"
        />
      </div>
      <div
        class="media__body"
      >
        <div
          class="track__name"
        >
          Is
        </div>
        <div>
          <a
            href="/artists/5VQE4WOzPu9h3HnGLuBoA6"
          >
            Vlasta Marek
          </a>
        </div>
      </div>
    </div>
  </td>
  <td>
    <a
      href="/albums/5Opp8ursJ6XoTTcqElvAJG"
    >
      Tibetan Bowls
    </a>
  </td>
  <td>
    Jan 15, 2015
  </td>
  <td
    align="right"
  >
    12:10
  </td>
</tr>
`;
